gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc

```diff
@@ -0,0 +1,153 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <string_view>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "gematria/datasets/bhive_importer.h"
+#include "gematria/datasets/find_accessed_addrs.h"
+#include "gematria/llvm/canonicalizer.h"
+#include "gematria/llvm/llvm_architecture_support.h"
+#include "gematria/utils/string.h"
+
+constexpr uint64_t kInitialRegVal = 10000;
+constexpr uint64_t kInitialMemVal = 2147483647;
+constexpr std::string_view kRegDefPrefix = "# LLVM-EXEGESIS-DEFREG ";
+constexpr std::string_view kMemDefPrefix = "# LLVM-EXEGESIS-MEM-DEF ";
+constexpr std::string_view kMemMapPrefix = "# LLVM-EXEGESIS-MEM-MAP ";
+constexpr std::string_view kMemNamePrefix = "MEM";
+
+namespace {
+unsigned int file_counter = 0;
+}
+
+ABSL_FLAG(std::string, bhive_csv, "", "Filename of the input BHive CSV file");
+ABSL_FLAG(
+    std::string, output_dir, "",
+    "Directory containing output files that can be executed by llvm-exegesis");
+
+int main(int argc, char* argv[]) {
+  absl::ParseCommandLine(argc, argv);
+
+  const std::string bhive_filename = absl::GetFlag(FLAGS_bhive_csv);
+  if (bhive_filename.empty()) {
+    std::cerr << "Error: --bhive_csv is required\n";
+    return 1;
+  }
+
+  const std::string output_dir = absl::GetFlag(FLAGS_output_dir);
+  if (output_dir.empty()) {
+    std::cerr << "Error: --output_dir is required\n";
+    return 1;
+  }
+
+  std::string register_defs_lines;
+  const std::unique_ptr<gematria::LlvmArchitectureSupport> llvm_support =
+      gematria::LlvmArchitectureSupport::X86_64();
+  const llvm::MCRegisterInfo& MRI = llvm_support->mc_register_info();
+  for (llvm::MCPhysReg I = 1, E = MRI.getNumRegs(); I != E; ++I) {
+    // Filter out subregisters. Only include the super register.
+    bool is_sub_reg = false;
+    for (auto SuperReg : MRI.superregs(I)) {
+      if (MRI.isSubRegister(SuperReg, I)) {
+        is_sub_reg = true;
+        break;
+      }
+    }
+    if (is_sub_reg) {
+      continue;
```

Comment:
This needs some adjustment. It returns all possible registers for X86 as it doesn't take into account the current CPU features. So you'll end up with all the vector registers and even GPRs that aren't on currently released processors (as APX is supported in upstream LLVM now).

It'll probably be a lot easier to just look through specific register classes.  `X86::GR64_NOREX2` and `X86::VR128` should get what you need. The following two loops shown an example of how to iterate over the classes to get the names.

```c++
for(unsigned i = 0; i < MRI.getRegClass(X86::GR64_NOREX2RegClassID).getNumRegs(); ++i) {
  std::cout << MRI.getName(MRI.getRegClass(X86::GR64_NOREX2RegClassID).getRegister(i)) << "\n";
}

for(unsigned i = 0; i < MRI.getRegClass(X86::VR128RegClassID).getNumRegs(); ++i) {
  std::cout << MRI.getName(MRI.getRegClass(X86::VR128RegClassID).getRegister(i)) << "\n";
}
```

At some point we'll probably want to wire this into a `TargetRegisterInfo` so we can actually query information on what vectors are available for the hardware we're running on, but for now just iterating through the classes should be fine.